### ISSUE TITLE
Type 'Record<keyof T, T[keyof T]>' is not assignable to type 'T'.

### ISSUE THREAD

**Issue by pelotom** [Microsoft/TypeScript#21515]
**TypeScript Version:**  2.7.1, 2.8.0-dev.20180131

**Search Terms:** 

"Type 'Record<keyof T, T[keyof T]>' is not assignable to type 'T'."

**Code**

```ts
function foo<T>(x: Record<keyof T, T[keyof T]>): T {
  return x;
}
```

**Expected behavior:**

Should type check, and did in version 2.6.2 and before.

**Actual behavior:**

```
Type 'Record<keyof T, T[keyof T]>' is not assignable to type 'T'.
  Type 'T[keyof T]' is not assignable to type 'T[P]'.
    Type 'keyof T' is not assignable to type 'P'.
```

**Comment by pelotom** [comment]
Never mind, I think this is actually correct, and TypeScript is catching something it didn't before.

```ts
const { y } = foo<{ x: number; y: boolean }>({ x: 3, y: 42 })
```

Here `y` would have type `boolean` but its runtime value would be a number.
